### DISCUSSION TITLE
PWM duty cycle linear output or solution for Burst firing AC

### DISCUSSION THREAD

**Discussion by radudorurus** [arendst/Tasmota#16444] [Q&A] [answered]
Hello.

I'd like to use PWM output on ESP32 for an AC power variator.

For this application I would need the PWM Dutycycle output to be identical with the command value (or the slider position).

I will use a separate device to convert dutycycle (or voltage) to burst fire command for AC power variator.

Is there a possibility to make the PWM dutycycle fit with the level/slider/command?
I mean "Dimmer 75" command to execute 75% PWM dutycycle on the configured output pin.

I found a workaround using Berry scripting for DAC outputs as voltage, still quite complicated.

I am open to other solutions for the application I mentioned (control a SSR-40 DA zero cross ON, zero cross OFF using burst fire)
My aplication is a power diverter and I wish to vary the power for the heater element in 10 steps of 10%.
For example:
- 20% power means 1 full wave ON (20ms), 4 full waves OFF (80ms)
- 50% power would be 1 full wave ON, 1 full wave OFF , 20ms on/ 20ms off.
- 30% would be 3 cycles 60ms ON, 7 cycles 140ms OFF

An OpenHab server (Blockly) will issue the power command directly based on some conditions (solar power, battery SOC, battery power, AC output power)
There are 3 heater elements of 1500W power, 2 of them being switched by relays and one with burst fire control for being able to combine all three and obtain any power 0-4500W in 150W step increment.

**Comment by sfromis** [comment, marked as answer]
One part of Tasmota PWM duty cycle handling is that default setup for PWM channels is to treat it as a light, meaning gamma correction to have a slider match the perceived brightness instead of numerical steps. You can turn this off using the `LedTable` command, or instead of using the lights support at all, you might want to use `SetOption15` for basic control via the `PWM` command setting duty cycle values 0..1023.

**Reply by radudorurus** [reply]
Thanks for the tips.
I will come back with feedback.

I was thinking for another solution of my application.
That would require an output PWM f=5Hz (a period of 200ms).
With this kind of signal I could command directly the SSR 40 DA.
But from the specification, ESP cannot output lower than 40Hz

**Reply by sfromis** [reply]
While ESP8266 PWM frequency is capped at minimum 40 Hz, on ESP32, Tasmota goes down to 2 Hz.

**Reply by radudorurus** [reply]
Thank you for the news.

I've read also this topic.
https://github.com/arendst/Tasmota/discussions/13123

If the frequency can be changed on the fly as low as 2Hz as you said, that makes possible to control directly the SSR by ESP32 without any intermediary hw.
I was thinking of an atTiny85, Arduino Nano, Pro mini or something similar with ADC read and one lookup table (switch - case set) where to set the ON and OFF times pairs for 10 or 20 cases. That would give me 1/10 (10%) power step or 1/20 (5%) step for the burst fire control.

**Reply by barbudor** [reply]
You mentioned zero crossing
If you need PWM synced on zero crossing for partial wave, this is AFAIK supported by Tasmota but never used it myself

**Reply by radudorurus** [reply]
There no need for syncing the PWM with AC wave.
The SSR has zero cross detection pwn turn ON and thecsame for turn OFF.

So, turning on the GPIO for 40ms will cause the SSR to fire ON when the next zerocross occurs.
When the GPIO turns off, the SSR stays ON until the next zerocross occurs.

**Reply by barbudor** [reply]
OK so you don't use partial wave. This is why you need low frequency.
Thanks for the detail

**Reply by radudorurus** [reply]
It works as expected.
I did some tests with 750W heating load.
I set the PWM frequency to 5, 10, 12.5, 20Hz.
5Hz - the power control step is minim 5% (half wave 10ms)
10Hz - the power control step is minim 10% (half wave 10ms)

See below an idea.

![Clipboard05](https://user-images.githubusercontent.com/17679049/189550198-779af737-96b6-48b1-9972-b2cd151643ca.jpg)


There are 2 variants working:
- use dimmer control with ledTable disabled and give command dimmer x (wanted percent duty cycle). Slider and toggle button exists in the interface.
- setoption15 OFF and use raw values of PWM. No slider and toggle button in the interface.

**Reply by dkebler** [reply]
@radudorurus    What was the `I did some tests with 750W heating load.` you used for your test?    I have Runtal wall heaters and for these type of heaters I have read that the typical control is a "long" cycle (e.g. 15 secs full power/voltage) instead of a fast duty discussed here.  I contacted the manufacturer and they were clueless and have never even tried "partial power with fast cycling" with their panels and thus could not recommend doing so. 

https://www.digitalhome.ca/threads/electronic-thermostats-and-baseboard-crackle.147705/
https://runtalnorthamerica.com/residential-electric-wall-panels/

**Reply by radudorurus** [reply]
I don't remember.
I know I had 1500W water heater element.
So one full wave ON and one full wave OFF results 750W.

**Comment by xjustbmw** [comment]
I am trying to do the same , what meter do you have? Can this method bypass any smart meter ?
Im trying to use the pv energy instead of sending it back to the grid, for almost free.
thanks

**Reply by radudorurus** [reply]
You can use any any powermeter that can read energy in both directions.
One of them is ShellyEM.

**Comment by xjustbmw** [comment]
I already have this data from my fronius smart meter, through the inverter.
I can read both directions , for each phase, via modbus tcp or even the fronius api.
The question is what is the best working method (preferably lan connected) to send the this power to the heater in order for it to work with newer smart operators meters. mine is AEG CST 0410. 

Also, for either small/large loads , its impossible using my fronius symo inverter to use all the available pv power, I end up usually sending some power to the grid on one phase, while using power from the grid on the other(s). Any way to combat this ?:)

thanks

**Reply by radudorurus** [reply]
If you do not want to diy, just buy Ohmpilot.

Otherwise, you need the smartmerer power value (positive or negative) in an automation server (openhab, homeassistant) which will issue commands to some tasmota clients (relays/ssr with loads ) based on the power value. (NodeRed or Blockly integrated in the automation server)

Maybe there is a way to get the Fronius inverter data into the automation server. There are a lot of "bridges" which does that for different devices.

I have 3 air conditioning units Mitsubishi connected to MELCloud server.
My setup with openhap has a software bridge between Melcloud and the interface so that I can see the status and activate any command from the Openhab server.

**Comment by xjustbmw** [comment]
Ohmpilot is not a bad solution.
Problem is that is expensive and it only modulates one phase. 
The other 2 are just simple relays that surely do not trigger in cloudy days/winter, exactly when I actually need this energy to the max,instead of pushing it to the grid.

I already have nodered running, querying these values from the inverter.I also have a few tasmota esps in the house. 

What type of signal should the tasmota (simple LED PWM , which frequency ?) send to which type of SSR in order to trick the operator ?
I keep reading about burst mode, phase control SSRs and I am not really sure which works on newer operator smart meters (if any)

any suggestions on how tasmota/ssr actually works in an installation ?

**Reply by radudorurus** [reply]
All right then.
Didn't you read the entire thread?

I warn you, this is not a solution for ongrid inverter.
Burst control for power variation on a diverter (ongrid setup) is not suitable because you have for 100ms time slot (lets say) 1kw power injection.
The diverter control will keep the 2kw load ON for 50ms and OFF for another 50ms.
The utility meter will record 1kw energy consumed for 50ms and 1kw energy injected for 50ms. It will not integrate the energy on such long time slot.
Ongrid diverters works with phase control, adjusting the power each 10ms.

There are standalone diverters which do this measuring in realtime, sensing zerocross, reading the voltage, the current and the direction of the  current (current transformer).
These diverters take decision very fast.
I have seen arduino uno/nano/promini doing this.
No doubt one ESP32 can do it. (2 analog inputs are needed for the current and for the voltage)

Anyway.
I use a zerocross ON, zerocross OFF ssr.
It is a SSR-40-DA (40A or 25A if you wish)

Frequency (ESP pwm) depends of what power resolution do you wish.

5Hz - total cycle is 200ms which gives you 20 half waves to control. That means a step o 5% of power of the load.
For example, if you need  35% power,  keep ssr ON for 71....79ms (duty cycle 36...39%)
If you choose 10hz frequency, you have 100ms cycle (10 halfwaves), 10% step resolution.

**Comment by xjustbmw** [comment]
Yes, I read all the post, but like you said it might not work for my ongrid setup. I am trying to emulate ongrid diverters functionality if possible.
Not really see much benefit of instant readings. Its not the end of the world with lets say 1 second readins from the inverter.

Can this work with tasmota + any type of ssr in my system work with the phase control, (10ms power adjustment) ?

Many thanks again for all the effort to explain this for newbies like me. 
I am sure there are others like me who might benefit from this in the long run.

**Reply by radudorurus** [reply]
You need to dig on tasmota documentation to find any refference to zero cross.
Otherwise you cannot perform phase control. The SSR must be another type (not the "DA") which fires anytime but autostops at zerocross.

**Comment by xjustbmw** [comment]
I see. that was the isssue then. Maybe a zercoross hardware SSR would do the trick ?

**Reply by radudorurus** [reply]
There are configurable SSR with selectable input as PWM, voltage (0-5v, 0-10V), current (4-20mA) and output as burst or phase.
Not cheap - 200-300eur

Having something like this, offers the possibility to control by Tasmota using PWM  (dutycycle and small capacitor on that output to integrate the signal) or voltage out 0-3.3V (berry scripting on DAC out ESP32)

**Comment by xjustbmw** [comment]
Alright, gets quite expensive but it might worth it in the long run, depending also on how long that SSR can work for..

**Comment by xjustbmw** [comment]
So , after a lot of searching I came across this:
https://blocnote360.wordpress.com/2022/03/20/routeur-solaire/

what do you think ?

**Comment by AlexDavidenko** [comment]
@radudorurus 
Looks like this is exactly the solution for my prototyping needs! Thanks for figuring this out, the math is very straight forward. I wasn’t aware of burst fire, was using phase angle control which has its quirks with Tasmota (namely that it overrides all the PWM channels even the ones that are not part of the “light”, so you can’t control
Anything else via PWM because the frequency is so low, AC control and LED control or even PWM fan control is out of the question when using the ZC driver). 
have you ever tested it on smaller loads like a regular incandescent/ halogen bulb?
I wonder how granular does it get and whether or not this is suitable for slow ramps (think 15 min long). It sound like with 2hz you can get to about 2% and I wonder if it’s good enough (you don’t see steps) for such a long ramp time before I go and buy bunch of SSRs. 
which heaters it’s obviously a none issue.

**Reply by dech1969** [reply]
Hi There was just reading your posts.
I’m relatively new to Home assistant just experimenting with 6 months.
I recently uninstalled some solar panels with no battery so I wanted to soak up excess solar with a home made water diverter controlled by home assistant. As opposed to using a my energi Eddi.
With the help of ai for coding I seem to have got it to work. I used and esp 32 and a 40 amp ssr the set up is controlled by PWM and  in 6 second cycle bursts. I have the SSR mounted on a heat sync. I use Shelly em as murphy sensors and have some nice coding for temperature control and when then cycles operate etc. I am just wondering how your system is going and has anyone else built something similar

**Reply by dech1969** [reply]
I tried using a robodyn 40 am dimmer first with a/c but I could not get that to work

**Reply by radudorurus** [reply]
> I tried using a robodyn 40 am dimmer first with a/c but I could not get that to work

I just tried once the robotdyn.
It worked for 20 minutes and something blew.
There was not too much amperage or the load too big.
I had only 1kw.

I abandoned the idea and chose the power control in steps of 500W.
I have 4 elements of 1kw and 4 smart relays.
Using 2 elements in series, give me 500W.
All automation is done in Home Assistant with helpers.
500 -1000 -1500 - 2000 - 2500 - 3000.

**Reply by dech1969** [reply]
Thanks for that this is a quick overview of mine 
Water Diverter Overview

Your water diverter setup is designed to prioritize using excess solar energy to heat your water while minimizing grid consumption. It works by controlling an SSR (solid-state relay) that powers a heating element in my cylinder based on live energy availability.
	1.	Live Export Measurement
	•	The system monitors your solar inverter’s output and your home’s real-time consumption using a Shelly em
	•	Live export is calculated as:
Live Export (W)} = Solar Generation} - 
	•	Positive values indicate surplus solar energy being exported to the grid.
	2.	SSR Control & Duty Cycle
	•	The SSR switches the immersion heater on/off rapidly, adjusting the duty cycle according to available export power.
	•	Example: If 50% of the target export is available, the SSR is ON 50% of the cycle and OFF 50%, effectively throttling the heater to match surplus energy.
	•	Thresholds are set to prevent the SSR from cycling unnecessarily at very low export levels (e.g., <10 W).
	3.	Temperature Cutoffs
	•	The PWM cycle respects cylinder temperature limits (e.g., stop heating at 53 °C, restart at 52 °C).
	•	This ensures the water doesn’t overheat while maximizing solar usage.
	4.	Practical Benefit
	•	This system ensures that surplus solar energy is used efficiently, reducing grid export and electricity costs while maintaining safe water temperatures.

**Reply by wjohn007** [reply]
I described my solution to this issue [here](https://github.com/arendst/Tasmota/discussions/19095), which is very similar to @radudorurus's.

The implementation was done in Berry. In addition to the heating element for hot water, the concept incorporates several infrared heaters.

John
